Add docs for OpenLDAP dynamic secrets (#10817)
hashicorp/vault: website/content/api-docs/secret/openldap.mdx

}
```

## Rotate Root Password

The `rotate-root` endpoint offers password rotation for the `binddn` entry used to manage OpenLDAP. This generated password will only be known to Vault and will not be retrievable once rotated.

| Method | Path                    |
| :----- | :---------------------- |
| `POST` | `/openldap/rotate-root` |

### Sample Request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    http://127.0.0.1:8200/v1/openldap/rotate-root
```

## Static Roles

> The `static-role` endpoint configures Vault to manage the passwords of existing individual OpenLDAP entries.

Reviewer: ```suggestion
The `static-role` endpoint configures Vault to manage the passwords of existing individual OpenLDAP or Active Directory domain user accounts.
```

Author: I'm hesitant to add this here as it really should be noted in the main docs that Active Directory is compatible. Otherwise we can pretty easily end up with a situation where we have to say `OpenLDAP (or Active Directory)` all over the API docs.

Reviewer: Depending on the config's schema, static roles also support AD right? Only suggesting this because the dynamic role section mentions AD, so this makes it sound like static role only supports OpenLDAP.

> This provides instructions to Vault on how to create an OpenLDAP or Active Directory domain user account.

Author: Updated to hopefully clarify this. I've removed the "or Active Directory" part of this line and tried to make the static & dynamic sections a bit more consistent in showing their support for AD.

### Parameters

- `dn` `(string: <required>)` - Distinguished name (DN) of entry Vault should manage.<br />
  **Example:** `cn=bob,ou=Users,dc=hashicorp,dc=com`

> - `rotation_period` `(string: <required>)` - How often Vault should rotate the password of the user entry, in seconds.

Reviewer 2: Maybe revise or remove `in seconds`, especially since one of the examples is "1h".

Author: Wordsmithed to be similar to the `default_ttl` and `max_ttl` variables.

  The minimum rotation period is 5 seconds.<br />
  **Example:** "5s", "1h".
- `username` `(string: <required>)` - The name of the user to be used when logging in. This is useful when `dn`

>   isn't used for login purposes (such as SSH).

Reviewer: FWIW you can also force a newline by adding an empty newline instead of using <br /> (though that is less clear and may be accidentally cleaned up later)

```suggestion
  isn't used for login purposes (such as SSH).<br />
```

Author: Adding an extra newline here doesn't add a `<br />`, it creates a new `<p>` tag which will have significantly more spacing:
<img width="900" alt="Screen Shot 2021-02-03 at 1 25 15 PM" src="https://user-images.githubusercontent.com/72365/106804943-4c01b180-6623-11eb-933c-ec230c1d3db7.png">
(the bottom one has a newline, while the top one has `<br />`)

Reviewer: Good point, `<br />` makes more sense then!

}
```

### Manually Rotate Static Role Password

> The `rotate-role` endpoint offers manual rotation of static roles.

Reviewer: ```suggestion
The `rotate-role` endpoint offers manual password rotation of static roles.
```

Author: Updated, though with different phrasing

| Method | Path                               |
| :----- | :--------------------------------- |
| `POST` | `/openldap/rotate-role/:role_name` |

### Sample Request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    http://127.0.0.1:8200/v1/openldap/rotate-role/:role_name
```

## Dynamic Roles

Create or update a dynamic role configuration. This provides instructions to Vault on how to create an OpenLDAP or
Active Directory domain user account.

### Create/Delete Dynamic Role Configuration

#### Parameters

| Method   | Path                       |
| :------- | :------------------------- |

> | `POST`   | `openldap/role/:role_name` |

Reviewer 2: ```suggestion
| `POST`   | `/openldap/role/:role_name` |
```

Author: :+1:

> | `DELETE` | `openldap/role/:role_name` |

Reviewer 2: ```suggestion
| `DELETE` | `/openldap/role/:role_name` |
```

Author: :+1:

    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \
    http://127.0.0.1:8200/v1/openldap/role/dynamic-role
```

### Read Dynamic Role Configuration

| Method | Path                       |
| ------ | -------------------------- |
| `GET`  | `openldap/role/:role_name` |

Retrieves a dynamic role's configuration.

#### Parameters

None

#### Response

`200 OK`

```json
{
  "creation_ldif": "...",
  "default_ttl": 1800,
  "deletion_ldif": "...",
  "max_ttl": 0,
  "rollback_ldif": "...",
  "username_template": "..."
}
```

### Templates

LDIF and username templates use the [Go template syntax](https://golang.org/pkg/text/template/) to construct the
LDIF/username that will be executed against the server. This is done because certain values (such as the username
and password within an LDIF statement) are not known at configuration time. Additionally, the template gives a
lot of flexibility to the operator to construct the LDIF and username.

Templated fields are delimited by `{{` and `}}`. To reference a field (such as the generated `Username`), a period is
placed in front of the field name. Example: `{{.Username}}`. This operates very similarly to

> [`jq`](https://stedolan.github.io/jq/). Spaces can be included between `{{` and `}}`. For instance:

Reviewer 2: Not sure about the `jq` reference, which might either: a) be unfamiliar  b) lead readers to think at other jq-ism will work here.

Author: That's a good point. I've removed the note about `jq`

may be: `v_test-role_1234567890`. Note the `-` (dash) in the `test-role`. If the LDAP system Vault is managing
restricts usernames/DNs to not allow for dashes (or other characters), the template must explicitly modify/remove those
characters. In this example, the template can be changed to `v_{{.RoleName | replace "-" "_"}}` to replace
the dashes with underscores. See [Template Functions](#template-functions) for more functions available.

`.RoleName` - The name of the role the credentials are being generated from.

`.DisplayName` - The display name associated with the user making the request against Value.

#### Template Functions

Both the LDIF templates as well as the username template support using the following functions:

`random` - Generates a random string from lowercase letters, uppercase letters, and numbers. Must include a
number indicating how many characters to generate.<br />
**Example:** `{{random 20}}` generates 20 random characters

`truncate` - truncates the previous value to the provided number of characters.<br />
**Example:** `{{.FieldName | truncate 10}}`

> `truncate_sha256` - Truncates the previous value to the provided number of characters. The last 8 characters of the

Reviewer 2: It might be helpful to describe when this function is useful.

Author: Updated to have more details. I also included a more concrete scenario example. Let me know how this reads.

truncated value will be the first 8 characters of the SHA256 hash of the truncated characters.<br />
**Example:** `{{.FieldName | truncate_sha256 20}}`. If `FieldName` is `abcdefghijklmnopqrstuvwxyz`, all characters after
the 12th (`l`) are removed and SHA256 hashed (`872808ffbf0158b200d397a9884342497f23b153b1630229014cea1886ca6f20`.
The first 8 characters of the hash (`872808ff`) are then appended to the end of the first 12 characters from the
original value: `abcdefghijkl872808ff`.

> `uppercase` - Uppercases the provided value.

Reviewer 2: Some of these end in `<br />`, but some don't?

Author: They all should to put the `Example` on the next line. Looks like I missed a couple =/

`.RoleName` - The name of the role the credentials are being generated from.

`.DisplayName` - The display name associated with the user making the request against Value.

#### Template Functions

Both the LDIF templates and the username template use the [Go template language](https://golang.org/pkg/text/template)
so all [functions](https://golang.org/pkg/text/template/#hdr-Functions) and capabilities from that language are
supported including functions such as `printf`.

In addition to the functions available through the template language, the following custom functions are also available:

`random` - Generates a random string from lowercase letters, uppercase letters, and numbers. Must include a
number indicating how many characters to generate.<br />
**Example:** `{{random 20}}` generates 20 random characters

`truncate` - truncates the previous value to the provided number of characters.<br />
**Example:** `{{.FieldName | truncate 10}}`

`truncate_sha256` - Truncates the previous value to the provided number of characters. The last 8 characters of the
truncated value will be the first 8 characters of the SHA256 hash of the truncated characters.

> **Example:** `{{.FieldName | truncate_sha256 20}}`.<br />
> If `FieldName` is `abcdefghijklmnopqrstuvwxyz`, all characters after the 12th (`l`) are removed and SHA256 hashed
> (`872808ffbf0158b200d397a9884342497f23b153b1630229014cea1886ca6f20`). The first 8 characters of the hash (`872808ff`)
> are then appended to the end of the first 12 characters from the original value: `abcdefghijkl872808ff`.
> 
> This can aid in identifying values (such as the role & display names) if they need to be truncated to a particular
> length, particularly if the value has common prefixes. An example of this is having many roles with a common prefix in
> the name of the role, but the role name is truncated such that only the prefix is shown. This function will keep the
> non-prefix part of the role name in the username which will aid in identification while also keeping uniqueness.
> 
> **Example Scenario:**<br />
> Two roles exist: `mylongprefix-foobar` and `mylongprefix-bazqux`. If the template is
> `v_{{.RoleName | truncate 12}}_{{unix_time}}` the username would be `v_mylongprefix_1234567890` for both of the roles.
> However, if the template is `v_{{.RoleName | truncate_sha256 12}}_{{unix_time}}` the username for the first role
> would be: `v_mylo998a317d_1234567890` and the username for the second role would be: `v_mylo224e847c_1234567890`.

Reviewer: The two examples can be combined into one to convey the reasoning.

```suggestion
This can aid in identifying values (such as the role & display names) if they need to be truncated to a particular
length, particularly if the value has common prefixes. An example of this is having many roles with a common prefix in
the name of the role, but the role name is truncated such that only the prefix is shown. This function will keep the
non-prefix part of the role name in the username which will aid in identification while also keeping uniqueness.

**Example:**<br />
`v_{{.RoleName | truncate_sha256 12}}_{{unix_time}}`. <br />
If two roles exist for this template, `mylongprefix-foobar` and `mylongprefix-bazqux`, the username for the first role
would be `v_mylo998a317d_1234567890` and the username for the second role would be `v_mylo224e847c_1234567890`.
```

Author: :+1: with some tiny formatting changes
